### ISSUE TITLE
Payee matching

### ISSUE THREAD

**Issue by vomikan** [moneymanagerex/moneymanagerex#3148]
Payee definition is simple in MMEx. Ability to add a default category is nice. Importing transactions from the bank is allowed in the product in the CSV, QIF formats. During the import, would like a payee matching feature, which is present in Kmymoney, attached screenshot. This would help the payee's definition more readable.

![image](https://user-images.githubusercontent.com/6836805/112721175-957bc800-8f13-11eb-8a93-87969dfff882.png)

https://forum.moneymanagerex.org/viewtopic.php?f=8&t=10362

**Comment by tactilis** [comment]
> During the import, would like a payee matching feature

I would find this enhancement a great time saver.

Below are some example transactions (minus the Paid In, Paid Out, and Balance columns) in a CSV file from my bank showing various strings in the Description column and the corresponding Payee that I would map to.

Based on these, I would need the option to perform a **case blind** match of the strings (shown in red):
- at the start  `(^string.*)`
- at the end `(.*string$)`
- in the middle `(.*string.*)`

![image](https://user-images.githubusercontent.com/4060092/201544796-a5316b3a-b484-4383-9942-d580929f5751.png)

**Comment by n-stein** [comment]
Been squashing bugs lately, but I think its time to work on another useful feature. I've been playing around with this, looking for some feedback.

I added an input to the bottom of the Edit Payee dialog for entering match patterns. 
![image](https://user-images.githubusercontent.com/16599929/221738661-30ebb91e-7cf0-4d46-932b-eecb56be391c.png)

Enter any RegEx you want. If the expression is invalid, you'll be given a warning and the text turns red:
![image](https://user-images.githubusercontent.com/16599929/221739422-a4aec011-b8b7-46bd-82e3-6b726482ecde.png)

Filling the bottom cell automatically adds another row
![image](https://user-images.githubusercontent.com/16599929/221738391-0a4fc0e4-d5b5-4552-b94a-7df58815f9bf.png)

Patterns are displayed in the payee dialog
![image](https://user-images.githubusercontent.com/16599929/221744197-333a6ef5-fc28-420a-baa4-c27d615f91a1.png)

I added a new checkbox on the import screen. Enabling this checkbox will match the payee across all regular expressions in the pattern field and replace with the first match. If no match is found it will create a new payee.
![image](https://user-images.githubusercontent.com/16599929/221744628-3b4c13b4-1e8b-49f0-b3f4-7eec8f8dbdeb.png)

Matched Payees. Category is also autofilled if one is specified in the payee dialog.
![image](https://user-images.githubusercontent.com/16599929/221746626-633f43ea-7610-4fb9-b065-9a577ee3f0bd.png)

This uses an additional text field PATTERN in table PAYEE_V1, so will require DB upgrade.

**Comment by n-stein** [comment]
> 1\. I'd maybe change the `Match Patterns` label to `Match Patterns on Import`... This will make the dialog slightly wider

As a single line

![image](https://user-images.githubusercontent.com/16599929/221928783-308b5deb-1679-4f12-aacb-90de0b5a7949.png)

We can wrap the `on Import` to a second line to keep the label column width from increasing. I think this looks a bit better, but I don't have a strong opinion on the matter.

![image](https://user-images.githubusercontent.com/16599929/221929144-fc60958d-b64f-4338-82db-d881377e8a0a.png)


> 2\. I assume that during import, patterns are checked in the order shown and the first that matches applies.

This is correct. The match is processed in the order shown. I left the row headers (1, 2, 3...) on the table to try to show there is an order of precedence. I can change the labels to `1st`, `2nd`, `3rd` etc if that is clearer. Note there is nothing preventing the users from entering overlapping regex in multiple payees, so it will pick whichever payee matches first. Default check order is by PAYEEID, but we can order by PAYEENAME if it should check alphabetically instead.

> I'd like to be able to reorder the patterns (using Move Up/Down buttons?)

Text?

![image](https://user-images.githubusercontent.com/16599929/221926197-728d844a-d407-4dc9-a918-4bc1b94099fb.png)

Arrows?

![image](https://user-images.githubusercontent.com/16599929/221928112-d2b36889-fd2c-4757-8bf2-896eef8e2554.png)


> 3.\ What happens if the user deletes a pattern?

The behavior is exactly as you describe, i.e. the list is compressed before saving and any blank rows are discarded so the next time it is opened the gaps are removed. The list doesn't auto shift everything up, although I suppose I could make it do that if it would be beneficial.


> 4\. Can you confirm that the matching is case blind?

It is indeed.


> 5\. Will the Payee matching apply to QIF as well as CSV/XML files? This is just a question; I currently do not use QIF.

Yes, I will apply the same to QIF import once we get CSV ironed out.

**Comment by tactilis** [comment]
> 1. We can wrap the `on Import` to a second line to keep the label column width from increasing. I think this looks a bit better

Agreed.
It's more in proportion to the pattern block.

>  2. I left the row headers (1, 2, 3...) on the table to try to show there is an order of precedence. I can change the labels to 1st, 2nd, 3rd etc if that is clearer.

I think 1, 2, 3... looks best.

> 2 Note there is nothing preventing the users from entering overlapping regex in multiple payees, so it will pick whichever payee matches first. Default check order is by PAYEEID, but we can order by PAYEENAME if it should check alphabetically instead.

I'd leave it as PAYEEID check order. The user should not rely on a specific implementation detail if they create overlapping regex in multiple payees.

For info, the KMyMoney Manual says: _"Note: If a transaction payee matches more than one Match Key, the behavior is undefined. KMyMoney will arbitrarily pick one of the matching payees"._  Of course, it won't be arbitrary, but that sort of warning allows the implementation to be changed if required at a later date.

> 2. Text? Arrows?  (for move up/down)

Arrows seem OK to me.
Will you be able to mode them according to the current row? i.e, click in the top row and only the down arrow is active, etc.

> 3.  the list is compressed before saving and any blank rows are discarded so the next time it is opened the gaps are removed. 

Ideal.

> The list doesn't auto shift everything up, although I suppose I could make it do that if it would be beneficial.

IMO, adds complexity and is not necessary.

**Comment by n-stein** [comment]
> Will you be able to mode them according to the current row? i.e, click in the top row and only the down arrow is active, etc.

I certainly can, but I'm not sure this is desired. As it works now the code checks if the row is already at the top/bottom and does nothing if it is already there. This allows you to select multiple rows and move them up all at once without worrying about the top/bottom row being part of the selection.

For example, you start with this pattern list

```
1. PatternA
2. PatternB
3. PatternC
4. PatternD
```

If you select `PatternB` and `PatternD` and click the up arrow, it moves each up one row:
```
1. PatternB
2. PatternA
3. PatternD
4. PatternC
```

If you don't change your selection and simply click the up arrow again, only `PatternD` moves because `PatternB` is already at the top. This would not be possible if we disable the up arrow, it would require you to deselect `PatternB` before clicking up again

```
1. PatternB
2. PatternD
3. PatternA
4. PatternC
```

**Comment by tactilis** [comment]
> I certainly can, but I'm not sure this is desired. As it works now the code checks if the row is already at the top/bottom and does nothing if it is already there. This allows you to select multiple rows and move them up all at once without worrying about the top/bottom row being part of the selection.

Ideal.

**Comment by n-stein** [comment]
> Would it be possible to add another checkbox as shown? It would only be active if `Pattern Match Payees' is ticked?

Sure, I will add this in.

**Comment by rmelillo76** [comment]
This is really slick! Any opinion on adding a "Test" button below the arrows and would present the user with a small dialogue to input some sample text to then validate against the pattern to indicate if it would be a match or not. Just a thought. Maybe useful, even though there are plenty of regex tools available online.

**Comment by n-stein** [comment]
Just a note: 

Right now the patterns are stored in the database with a separator of `+++` like `PatternA+++PatternB+++PatternC`. Since the patterns can contain any regex, the separator has to be something explicitly not allowed in regex for delimiting purposes.  

If anyone has any other suggestions that would work better, let me know.

**Comment by tactilis** [comment]
> > Would it be possible to add another checkbox as shown? It would only be active if `Pattern Match Payees' is ticked?
> 
> Sure, I will add this in.

Thank you.

Contrary string to what I suggested above, could the string added to Notes be changed from:

Payee _[pattern_string]_ matched _[original_payee_text_from_file]_

to:

_[original_payee_text_from_file]_ matched by  _[pattern_string]_

because:
1. It's shorter
2. It places the _[original_payee_text_from_file]_ first which enables sorting by the Notes column to group transactions having the same source string.

BTW, I don't intend the [ and ] character to be included in the Notes. You likely understood that though 😉

**Comment by tactilis** [comment]
@n-stein
> Right now the patterns are stored in the database with a separator of `+++`
> Maybe json_writer would be better.

JSON feels more 'standard' for storage than using a specific separator.

**Comment by tactilis** [comment]
> Should the actual note and this appended string be separated by anything other than a space?

Yes. A new line, if any Note text is imported from the file.

Two examples:

CSV File Note: 
_(no text)_

Transaction Notes:
`SomeStore #3256 REFID25784 matched by SomeStore`

<br>

CSV File Note: 
Some note text here

Transaction Notes:
`Some note text here`
`SomeStore #3256 REFID25784 matched by SomeStore`

<br>

I realise that _appending_ the match details to any note text means that I won't be able to group all pattern matched transactions by sorting on the Notes column, but I think appending is preferable to prepending, and in any case I can use the Transaction Filter to easily find pattern matched payees as shown:

![image](https://user-images.githubusercontent.com/4060092/222163960-53432d14-575a-4ea8-a46c-46ab4b49c7d3.png)

**Comment by n-stein** [comment]
> you may wish to consider whether it would be preferable to make your patterns work like the Transaction Filter to give the option of simple matching for non-regex savvy users as well as providing `regex:`

Done. Specifying `regex:` will use regex for the match and checks the pattern for validity. Without the prefix it functions using the normal ?/* wildcard match just like the notes filter.
![image](https://user-images.githubusercontent.com/16599929/222544018-41673ae1-66a3-494f-9e69-c1448bffaddb.png)

**Comment by n-stein** [comment]
Once https://github.com/moneymanagerex/database/pull/47 is merged I will update the submodule and push this for testing.

**Comment by tactilis** [comment]
The tooltip text will need some small amendments:
1. The grammar can be improved.
2. The text contains a factual error.

I am in the process of documenting some inconsistencies in the way we use search string wildcards in different places in MMEX. It will propose some changes to the text of the associated tooltips.

I will raise a separate issue (hopefully tomorrow) and then edit this comment to make reference to it.

Any changes that arise from it won't affect the testing of your payee matching functionality.

**Comment by n-stein** [comment]
QIF import done. Added the same checkboxes
![image](https://user-images.githubusercontent.com/16599929/222603675-486e5dd5-9c0a-4d73-bc2a-da80cf364fdf.png)

Checking the match box updates the preview window to show which payees are matched by a pattern
![image](https://user-images.githubusercontent.com/16599929/222603859-ef42dd6d-fd5b-4cd0-b126-d3ca81adb3f6.png)

**Comment by tactilis** [comment]
> QIF import done. Added the same checkboxes

I just noticed that for QIF, there is already an `Include payee field in notes` checkbox.

Not being a QIF user, I'd never noticed that before. It seems that it achieves what I had previously been doing when importing CSV by manually duplicating the payee data in the file and mapping the 2nd instance to Notes.

As its function is related to `Add match details to Notes`, I'm wondering whether we could rationalize import by linking three checkboxes and rewording the labels for QIF, _and_ similarly for CSV/XML: 

```
  1. [ ] Pattern match Payees

  2. [ ] Add Payee field to Notes
  3. [ ] Include pattern match details in Notes
```
  
Box 1 would be always available
Box 2 would be always available
Box 3 would be greyed out unless both Box 1 _and_ 2 were checked.

<br>

> Checking the match box updates the preview window to show which payees are matched by a pattern

Nice. You could argue that having a preview, which can be looked at prior to import, reduces the need to `Include pattern match details in Notes`. However, I think it's still useful to have this ability - and it's needed for CSV/XML import which doesn't have a preview function.

**Comment by n-stein** [comment]
> it's needed for CSV/XML import which doesn't have a preview function.

There is a preview in CSV import, but it just shows the payee text from the file. Would it be helpful to include match details in that preview?

**Comment by tactilis** [comment]
> There is a preview in CSV import

It's only a _partial_ preview. It shows (up to) the first 50 and last 50 lines in the file:

![image](https://user-images.githubusercontent.com/4060092/222730586-374ab3f3-927d-479e-a6ce-ac50e3aacab8.png)

It's purpose is to:

1. Allow the user to visualise the mapping of the data to MMEX fields.

2. Allow the user to correctly set `Rows to ignore: From start / From end` to skip preamble/postamble rows that some banks add. The ignored rows are greyed out in the preview.

Please see https://github.com/moneymanagerex/moneymanagerex/issues/3487 and the linked forum thread for some changes that were made to the preview to increase the number of rows displayed.

**Comment by n-stein** [comment]
Ah, thanks for that. I didn't realize there was a set row limit on the preview. Will leave it as-is then.

**Comment by tactilis** [comment]
>  I didn't realize there was a set row limit on the preview

I think there may be a case for removing that restriction and showing all rows now (see the discussion between @vomikan and me in  #3487). I never saw a significant performance problem when using 1.3.6 which did display all rows.

If anything is to be done about this, it should be in a separate issue from this one on Payee matching. After that, it would be helpful to have match details in the preview.

**Comment by n-stein** [comment]
What do you think about adding a tabbed preview similar to the QIF import has that shows details of categories and payees?

**Comment by tactilis** [comment]
> What do you think about adding a tabbed preview similar to the QIF import

I think that would be fantastic.

**Comment by n-stein** [comment]
In conjunction with the tabbed preview, perhaps adding a button to open the payee dialog from the import dialog would be useful to edit patterns prior to load. If you notice a payee is not captured properly by your patterns in the "Payee" tab you can easily open & edit the patterns rather having to close the import dialog, open/edit/close the payee dialog, then reopen the import dialog.

**Comment by tactilis** [comment]
> However, I don't want to clutter/overload the dialog if this would not be useful.

Banish that thought!

Yes, that would be very useful. If you could quickly see a list of source payee strings and their matches, that would allow you address each non-match straight away prior to doing the actual import.

**Comment by n-stein** [comment]
`Data` tab shows the same preview as before
![image](https://user-images.githubusercontent.com/16599929/222814195-ad199ac7-7fc2-4553-b8b3-1c75f0757c62.png)

`Payee` tab shows a summarized list of payees and how they will be handled
without matching:
![image](https://user-images.githubusercontent.com/16599929/222813954-fab63923-68a7-4d39-83c9-cc4ab434a1e8.png)

with matching
![image](https://user-images.githubusercontent.com/16599929/222814026-42091a24-c018-44c9-9cdb-a9ee47b502cc.png)

`Category` tab shows a summarized list of categories in the file
![image](https://user-images.githubusercontent.com/16599929/222814238-d71ad147-8d93-4f11-9d45-14ee86622110.png)

**Comment by n-stein** [comment]
> Yes, that would be very useful. If you could quickly see a list of source payee strings and their matches, that would allow you address each non-match straight away prior to doing the actual import.

I've made double-clicking on the tabs open their respective dialogs. Thoughts?

**Comment by tactilis** [comment]
> I've made double-clicking on the tabs open their respective dialogs. Thoughts?

Witchcraft / Sorcery / Devil worship!

It's not conventional to double click on a tab to invoke other functionality. So it's a rather hidden UI action that few will discover.

Unless, you mean double clicking on a payee/category row (maybe with a right click menu as a clue too)?

**Comment by n-stein** [comment]
> Unless, you mean double clicking on a payee/category row (maybe with a right click menu as a clue too)?

I did mean double clicking on the tab itself. I'll bind it to the table instead and add a right click menu with `Organize Payees...` and `Organize Categories...` on their respective tabs.

**Comment by tactilis** [comment]
> I'll bind it to the table instead and add a right click menu with `Organize Payees...` and `Organize Categories...` on their respective tabs.

That's great.

Now you are feeding off the rows, how feasible would be to invoke the Organize dialogs scrolled to the relevant entry? Or the nearest position alphabetically in the case of a missing item? This would be a big time saver if it is possible.

**Comment by n-stein** [comment]
> how feasible would be to invoke the Organize dialogs scrolled to the relevant entry? Or the nearest position alphabetically in the case of a missing item?

Done. It automatically opens to the target payee/category, or if there is no target identified it highlights the payee/category which comes prior alphabetically (by full name in the case of category). Here's a quick demo:

https://user-images.githubusercontent.com/16599929/223145183-0efc32fb-3962-43d2-98c8-8c4db7b1e1af.mp4

**Comment by n-stein** [comment]
AppVeyor build 402 can be tested now, let me know what you think.

**Comment by whalley** [comment]
On MacOS the unicode up/down characters don't render correctly, don't know why.
![CleanShot 2023-03-11 at 18 19 09](https://user-images.githubusercontent.com/487345/224505146-1909d496-f904-46a4-aeb3-b83c1184b063.png) 

Maybe we should actually be using a wxBitmapButton and the UPARROW/DOWNARROW images already in the theme files for this anyway?

**Comment by tactilis** [comment]
> On MacOS the unicode up/down characters don't render correctly

The 1, 2, 3, 4 fields are overly wide too

**Comment by n-stein** [comment]
> The 1, 2, 3, 4 fields are overly wide too

This is odd because I explicitly set the width in the code which works fine on Windows.

https://github.com/moneymanagerex/moneymanagerex/blob/master/src/payeedialog.cpp#L128

**Comment by vomikan** [comment]
You just have to try other codes.
https://stackoverflow.com/questions/13973106/unicode-ascii-arrows-in-8-directions
https://symbl.cc/en/collections/arrow-symbols/

**Comment by n-stein** [comment]
> Maybe we should actually be using a wxBitmapButton and the UPARROW/DOWNARROW images already in the theme files for this anyway?

The bitmap buttons look good.
![image](https://user-images.githubusercontent.com/16599929/225094382-6d698b3c-307b-4262-8d76-56f3e0d9743d.png)

> The 1, 2, 3, 4 fields are overly wide too

I switched to using wxGRID_AUTOSIZE for row labels. @whalley does that make them render smaller on macOS?

**Comment by whalley** [comment]
> I switched to using wxGRID_AUTOSIZE for row labels. @whalley does that make them render smaller on macOS?

Away at present, will test when I get back next week.

**Comment by n-stein** [comment]
RE: the 100 line preview limit, this is also currently limiting the list of payees/categories read from the file. I need to make a small change so that the 100 rows are shown in the preview, but all payees and categories are checked.

**Comment by tactilis** [comment]
> RE: the 100 line preview limit

But is a limit or 100 rows still necessary?

Just wondering what is the performance hit of:  for example 100 vs 200 vs 400 rows?

Is there an unacceptable slowdown beyond 100 lines?

**Comment by n-stein** [comment]
Some (not particularly scientific) performance testing... take this with a grain of salt.

I built a file using all available columns once (15 columns) with completely randomized unique strings for payees, categories, subcategories, notes ... basically the worst case scenario. Here are some results:

Using a file that has 1k rows:
```
Preview        Load Time
100 rows      ~0.7 seconds
200 rows      ~1.1 seconds
500 rows      ~1.9 seconds
1000 rows     ~3.7 seconds
```

Using a file that has 10k rows:
```
Preview        Load Time
100 rows      ~3.8 seconds
200 rows      ~4.0 seconds
500 rows      ~4.8 seconds
1000 rows     ~6.5 seconds
5000 rows     ~25.4 seconds
10000 rows    ~67.6 seconds
```

So based on my very limited set of data, I would say the length of the CSV file matters more than the preview row count for the first 500-1000 rows, after that the preview generation is a large performance hit. We could probably increase the preview to 500 rows without users even noticing. Or we can remove it entirely -- how many users are going to be loading 1k+ row CSV files? I'd guess not very many, and those that do won't be doing it regularly.

**Comment by tactilis** [comment]
> Or we can remove it entirely -- how many users are going to be loading 1k+ row CSV files? I'd guess not very many, and those that do won't be doing it regularly.

Agreed.

When I first started using MMEX, I did a one-off import of 2 years worth of transactions for each of my bank accounts. One of the accounts had around 1500 rows in the CSV.

~5 seconds load/preview is miniscule compared to the total time I spent obtaining the CSV in the first place, then manually payee matching & categorizing the transactions after import.

I could easily have divided the 2-year file into four 6-month chunks if performance had been an issue.

Thereafter, I load CSVs on an approx monthly basis (or every couple of months if I'm lazy), so the number of transactions is small.

<br>

Out of interest, what was the PC spec that you did your tests on:
- Processor type & speed?
- HDD or SSD?

**Comment by n-stein** [comment]
> Out of interest, what was the PC spec that you did your tests on:\n\nProcessor type & speed?\nHDD or SSD?

Im running:
* An AMD Phenom II 1090T from 2012
* 16GB DDR3 RAM (was using 8GB until last week when I got fed up with Visual Studio complaining about running out of heap space)
* A small SSD.

I suspect the SSD vs HDD wouldn't make much difference as the processing is all done in memory after the data is pulled from disk.

**Comment by whalley** [comment]
> > I switched to using wxGRID_AUTOSIZE for row labels. @whalley does that make them render smaller on macOS?
> 
> Away at present, will test when I get back next week.

Looks fine on MacOS... But I get a Debug Check failure...

![CleanShot 2023-03-20 at 19 36 53](https://user-images.githubusercontent.com/487345/226447812-5ceb7a92-51a5-4e34-8e41-6b1113e3fe4f.png)

Suspect you may need to guard against nothing being selected when first called.

**Comment by whalley** [comment]
When opening the Payee dialog through menu or toolbar icon, not tried through CSV import.

**Comment by whalley** [comment]
Also, get a crash when attempting to export CSV.

![CleanShot 2023-03-21 at 08 32 15](https://user-images.githubusercontent.com/487345/226553229-9f10fd20-5eb2-40d4-8f0e-4913968ba690.png)

There is a call to `mmUnivCSVDialog::compilePayeeRegEx()` which accesses `payeeMatchCheckBox_` when uninitialised.

Should this only be done only for Import? As the `payeeMatchCheckBox_` is only initialised in `CreateControls` for import

**Comment by n-stein** [comment]
> Of course only when wxWidgets and MMEX are built with Debug turned on.... Occurs when opening the Payee dialog through menu or toolbar icon, not tried through CSV import. Pressing continue opens dialog fine.

Hmm. I'm using the Debug libraries, but still can't get this issue to trigger on Windows. Regardless, I added a check that at least one item is selected to prevent calling EnsureVisible on a -1 index. I think that should fix it.


> Should this only be done only for Import? As the `payeeMatchCheckBox_` is only initialised in `CreateControls` for import

Absolutely should. I moved the call below the IsImporter check.

**Comment by tactilis** [comment]
Just mentioning that issue #1789 from Aug 2018 will be able to be closed when #3148 is released.

I don't think that #1789 has been referenced previously.

**Comment by n-stein** [comment]
Fixed an issue where editing a payee didn't completely refresh the match data causing old patterns/names to be used.
